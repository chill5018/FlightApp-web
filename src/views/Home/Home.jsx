import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { Input, Button } from 'reactstrap';
import Select from 'react-select';

import './index.scss';
import { validateForm } from '../../utils/helpers';

const options = [
  { label: 'Copenhagen', value: 'CPH' },
  { label: 'New York City (JFK)', value: 'JFK' },
];

const optionStyles = {
  control: styles => ({ ...styles, backgroundColor: 'white' }),
  option: (styles) => {
    const color = '#000';
    return {
      ...styles,
      color,
    };
  },
};

/* eslint-disable jsx-a11y/label-has-for */
class Home extends Component {
  constructor(props) {
    super(props);
    this.state = {
      departureCity: options[0],
      arrivalCity: options[1],
      departureDate: '2019-02-01',
      returnDate: '2019-02-14',
      isRoundtrip: true,
      ticketQty: 1,
    };

    this.handleChange = this.handleChange.bind(this);
    this.handleDepartureChange = this.handleDepartureChange.bind(this);
    this.handleArrivalChange = this.handleArrivalChange.bind(this);
    this.searchFlights = this.searchFlights.bind(this);
  }

  componentWillReceiveProps(newProps) {
    const { history } = this.props;
    if (newProps.flights) {
      history.push('/flights');
    }
  }

  handleChange = (event) => {
    /* eslint-disable prefer-destructuring */
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;

    this.setState({
      [name]: value,
    });
  }

  handleArrivalChange = (selectedOption) => {
    this.setState({ arrivalCity: selectedOption });
  }

  handleDepartureChange = (selectedOption) => {
    this.setState({ departureCity: selectedOption });
  }

  async searchFlights() {
    const { searchFlights } = this.props;
    const {
      departureCity,
      arrivalCity,
      departureDate,
      isRoundtrip,
      returnDate,
      ticketQty,
    } = this.state;

    const errors = validateForm(this.state);

    if (errors.length > 0) {
      this.setState({ errors });
    } else {
      /* eslint-disable max-len */
      await searchFlights(departureCity.value, arrivalCity.value, departureDate, isRoundtrip ? returnDate : undefined, ticketQty);
    }
  }

  render() {
    const {
      isRoundtrip,
      arrivalCity,
      departureCity,
      departureDate,
      errors,
      returnDate,
    } = this.state;
    return (
      <div className="app-main">
        <div className="app-main-wrapper">
          <h4 className="app-main-h4">
            Book a flight
          </h4>
          <form onSubmit={() => {}} className="app-main-form app-main-col_1">
            <div className="app-main-col_2 dropdown-container">
              { errors && errors.map(error => (
                <li>
                {error}
                </li>
              ))}
            </div>
            <div className="app-main-col_2 dropdown-container">
              <label className="app-main-form-label">
                From
              </label>
              <Select
                className="app-main-form-input"
                name="departureCity"
                onChange={this.handleDepartureChange}
                options={options}
                placeholder="Departure City"
                styles={optionStyles}
                value={departureCity}
              />
            </div>

            <div className="app-main-col_2 dropdown-container">
              <label className="app-main-form-label">
                To
              </label>
              <Select
                className="app-main-form-input"
                name="arrivalCity"
                onChange={this.handleArrivalChange}
                options={options}
                placeholder="Arrival City"
                styles={optionStyles}
                value={arrivalCity}
              />
            </div>

            <div className="app-main-col_2">
              <div className="app-main-col app-main-col_2">
                  <input
                    checked={isRoundtrip}
                    className="app-main-form-checkbox"
                    id="roundtrip"
                    name="isRoundtrip"
                    onChange={this.handleChange}
                    type="checkbox"
                  />
                <label className="app-main-form-label" htmlFor="roundtrip">
                  Round Trip
                </label>
              </div>

              <div className="app-main-col app-main-col_2">
                <input
                  type="number"
                  className="app-main-form-number"
                  placeholder="1"
                  name="ticketQty"
                  onChange={this.handleChange}
                />
                <label className="app-main-form-label" htmlFor="roundtrip">
                  Traveller(s)
                </label>
              </div>

            </div>

            <div className="app-main-col_2">
              <div className="app-main-col app-main-col_2">
                <label className="app-main-form-label">
                  Depart
                </label>
                <br />
                <Input
                  type="date"
                  className="app-main-form-date_picker"
                  name="departureDate"
                  value={departureDate}
                  onChange={this.handleChange}
                />
              </div>

              { isRoundtrip && (
                <div className="app-main-col app-main-col_2">
                  <label className="app-main-form-label">
                    Return
                  </label>
                  <br />
                  <Input
                    type="date"
                    className="app-main-form-date_picker"
                    name="returnDate"
                    value={returnDate}
                    onChange={this.handleChange}
                  />
                </div>
              )}
            </div>

            <div className="app-main-form-button">
              <Button
                className="w-100 mt-5 app-main-form-button_color"
                onClick={this.searchFlights}
              >
                Search
              </Button>
            </div>
          </form>

            <hr className="app-main-hr" />

          <h4 className="app-main-h4">
            View a booking
          </h4>
          <form onSubmit={() => {}} className="app-main-form app-main-col_1">
              <div className="app-main-col_2">
                <Input className="app-main-form-input" placeholder="Booking reference" />
              </div>

            <div className="app-main-form-button">
              <Button className="w-100 mt-5 app-main-form-button_color">
                Search
              </Button>
            </div>
          </form>
        </div>
      </div>
    );
  }
}

export default withRouter(Home);
