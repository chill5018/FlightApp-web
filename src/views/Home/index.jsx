import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { searchFlights } from '../../reducers/flights';
import { getBooking } from '../../reducers/bookings';
import Home from './Home';

/* eslint-disable arrow-body-style */
const mapStateToProps = (state) => {
  return {
    flights: state.flights.flights,
    loading: state.flights.loading,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    ...bindActionCreators({
      searchFlights,
      getBooking,
    }, dispatch),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Home);
