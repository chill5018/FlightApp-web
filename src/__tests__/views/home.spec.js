import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';

import Home from '../../views/Home';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('Home Snapshot', () => {
  it('renders correctly', () => {
    const initialState = {
      flights: {
        flights: undefined,
      },
    };

    const component = renderer.create(
    <Provider store={mockStore(initialState)}>
      <BrowserRouter>
        <Home />
      </BrowserRouter>
    </Provider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
