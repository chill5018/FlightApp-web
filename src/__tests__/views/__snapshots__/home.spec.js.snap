// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Snapshot renders correctly 1`] = `
<div
  className="app-main"
>
  <div
    className="app-main-wrapper"
  >
    <h4
      className="app-main-h4"
    >
      Book a flight
    </h4>
    <form
      className="app-main-form app-main-col_1"
      onSubmit={[Function]}
    >
      <div
        className="app-main-col_2 dropdown-container"
      >
        <label
          className="app-main-form-label"
        >
          From
        </label>
        <div
          className="css-10nd86i app-main-form-input"
          onKeyDown={[Function]}
        >
          <div
            className="css-1ns25le"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="css-1hwfws3"
            >
              <div
                className="css-1492t68"
              >
                Departure City
              </div>
              <div
                className="css-1g6gooi"
              >
                <div
                  className=""
                  style={
                    Object {
                      "display": "inline-block",
                    }
                  }
                >
                  <input
                    aria-autocomplete="list"
                    autoCapitalize="none"
                    autoComplete="off"
                    autoCorrect="off"
                    disabled={false}
                    id="react-select-2-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    spellCheck="false"
                    style={
                      Object {
                        "background": 0,
                        "border": 0,
                        "boxSizing": "content-box",
                        "color": "inherit",
                        "fontSize": "inherit",
                        "opacity": 1,
                        "outline": 0,
                        "padding": 0,
                        "width": "1px",
                      }
                    }
                    tabIndex="0"
                    type="text"
                    value=""
                  />
                  <div
                    style={
                      Object {
                        "height": 0,
                        "left": 0,
                        "overflow": "scroll",
                        "position": "absolute",
                        "top": 0,
                        "visibility": "hidden",
                        "whiteSpace": "pre",
                      }
                    }
                  >
                    
                  </div>
                </div>
              </div>
            </div>
            <div
              className="css-1wy0on6"
            >
              <span
                className="css-d8oujb"
              />
              <div
                aria-hidden="true"
                className="css-1ep9fjw"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-19bqh2r"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <input
            name="departureCity"
            type="hidden"
            value=""
          />
        </div>
      </div>
      <div
        className="app-main-col_2 dropdown-container"
      >
        <label
          className="app-main-form-label"
        >
          To
        </label>
        <div
          className="css-10nd86i app-main-form-input"
          onKeyDown={[Function]}
        >
          <div
            className="css-1ns25le"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className="css-1hwfws3"
            >
              <div
                className="css-1492t68"
              >
                Arrival City
              </div>
              <div
                className="css-1g6gooi"
              >
                <div
                  className=""
                  style={
                    Object {
                      "display": "inline-block",
                    }
                  }
                >
                  <input
                    aria-autocomplete="list"
                    autoCapitalize="none"
                    autoComplete="off"
                    autoCorrect="off"
                    disabled={false}
                    id="react-select-3-input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    spellCheck="false"
                    style={
                      Object {
                        "background": 0,
                        "border": 0,
                        "boxSizing": "content-box",
                        "color": "inherit",
                        "fontSize": "inherit",
                        "opacity": 1,
                        "outline": 0,
                        "padding": 0,
                        "width": "1px",
                      }
                    }
                    tabIndex="0"
                    type="text"
                    value=""
                  />
                  <div
                    style={
                      Object {
                        "height": 0,
                        "left": 0,
                        "overflow": "scroll",
                        "position": "absolute",
                        "top": 0,
                        "visibility": "hidden",
                        "whiteSpace": "pre",
                      }
                    }
                  >
                    
                  </div>
                </div>
              </div>
            </div>
            <div
              className="css-1wy0on6"
            >
              <span
                className="css-d8oujb"
              />
              <div
                aria-hidden="true"
                className="css-1ep9fjw"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <svg
                  aria-hidden="true"
                  className="css-19bqh2r"
                  focusable="false"
                  height={20}
                  viewBox="0 0 20 20"
                  width={20}
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <input
            name="arrivalCity"
            type="hidden"
            value=""
          />
        </div>
      </div>
      <div
        className="app-main-col_2"
      >
        <div
          className="app-main-col app-main-col_2"
        >
          <input
            checked={true}
            className="app-main-form-checkbox"
            id="roundtrip"
            name="isRoundtrip"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="app-main-form-label"
            htmlFor="roundtrip"
          >
            Round Trip
          </label>
        </div>
        <div
          className="app-main-col app-main-col_2"
        >
          <input
            className="app-main-form-number"
            name="ticketQty"
            onChange={[Function]}
            placeholder="1"
            type="number"
          />
          <label
            className="app-main-form-label"
            htmlFor="roundtrip"
          >
            Traveller(s)
          </label>
        </div>
      </div>
      <div
        className="app-main-col_2"
      >
        <div
          className="app-main-col app-main-col_2"
        >
          <label
            className="app-main-form-label"
          >
            Depart
          </label>
          <br />
          <input
            className="app-main-form-date_picker form-control"
            name="departureDate"
            onChange={[Function]}
            type="date"
          />
        </div>
        <div
          className="app-main-col app-main-col_2"
        >
          <label
            className="app-main-form-label"
          >
            Return
          </label>
          <br />
          <input
            className="app-main-form-date_picker form-control"
            name="returnDate"
            onChange={[Function]}
            type="date"
          />
        </div>
      </div>
      <div
        className="app-main-form-button"
      >
        <button
          aria-label={null}
          className="w-100 mt-5 app-main-form-button_color btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          Search
        </button>
      </div>
    </form>
    <hr
      className="app-main-hr"
    />
    <h4
      className="app-main-h4"
    >
      View a booking
    </h4>
    <form
      className="app-main-form app-main-col_1"
      onSubmit={[Function]}
    >
      <div
        className="app-main-col_2"
      >
        <input
          className="app-main-form-input form-control"
          placeholder="Booking reference"
          type="text"
        />
      </div>
      <div
        className="app-main-form-button"
      >
        <button
          aria-label={null}
          className="w-100 mt-5 app-main-form-button_color btn btn-secondary"
          onClick={[Function]}
        >
          Search
        </button>
      </div>
    </form>
  </div>
</div>
`;
